### PR TITLE
Add ABCI++ methods: PrepareProposal and ProcessProposal

### PR DESCRIPTION
**Pull request by Manav-Aggarwal**
Add ABCI++ methods: PrepareProposal and ProcessProposal

### CODE REVIEW

**Review by tzdybal — APPROVED** [review]
I assume the changes are "imported" from some other branch in upstream - is it correct?
Could you include this information in comment / commit message?

File: abci/client/client.go [hunk 1]
```diff
@@ -34,6 +34,7 @@ type Client interface {
 	QueryAsync(types.RequestQuery) *ReqRes
 	CommitAsync() *ReqRes
 	InitChainAsync(types.RequestInitChain) *ReqRes
+	PrepareProposalAsync(types.RequestPrepareProposal) *ReqRes
```
Comment (evan-forbes) [inline abci/client/client.go:37]: are we even using async methods? why not have process here as well?

File: abci/client/socket_client.go [hunk 2]
```diff
@@ -279,6 +279,14 @@ func (cli *socketClient) ApplySnapshotChunkAsync(req types.RequestApplySnapshotC
 	return cli.queueRequest(types.ToRequestApplySnapshotChunk(req))
 }
 
+func (cli *socketClient) PrepareProposalAsync(req types.RequestPrepareProposal) *ReqRes {
+	return cli.queueRequest(types.ToRequestPrepareProposal(req))
+}
+
+func (cli *socketClient) ProcessProposalAsync(req types.RequestProcessProposal) *ReqRes {
+	return cli.queueRequest(types.ToRequestProcessProposal(req))
+}
```
Comment (evan-forbes) [inline abci/client/socket_client.go:288]: why is this not in the interface above?

**Review by evan-forbes — APPROVED** [review]
I'm not 100% sure what needs to be added for rollmint, but this looks reasonable to me.

Left one potentially blocking question

**Comment by evan-forbes** [comment]
is the failing test flaky? I can't rerun it

I'm assuming its just flaky due to the message, so I approved, but we should obvi get that passing before we can merge

**Pull request closed without merging**